code/game/machinery/vendors/vending.dm

```diff
@@ -605,95 +605,98 @@
-				return
-			var/datum/data/vending_product/R = display_records[key]
-			if(!istype(R))
-				to_chat(user, "<span class='warning'>ERROR: unknown vending_product record. Report this bug.</span>")
-				return
-			var/list/record_to_check = product_records + coin_records
-			if(extended_inventory)
-				record_to_check = product_records + coin_records + hidden_records
-			if(!R || !istype(R) || !R.product_path)
-				to_chat(user, "<span class='warning'>ERROR: unknown product record. Report this bug.</span>")
-				return
-			if(R in hidden_records)
-				if(!extended_inventory)
-					// Exploit prevention, stop the user purchasing hidden stuff if they haven't hacked the machine.
-					to_chat(user, "<span class='warning'>ERROR: machine does not allow extended_inventory in current state. Report this bug.</span>")
-					return
-			else if(!(R in record_to_check))
-				// Exploit prevention, stop the user
-				message_admins("Vending machine exploit attempted by [ADMIN_LOOKUPFLW(user)]!")
-				return
-			if(R.amount <= 0)
-				to_chat(user, "Sold out of [R.name].")
-				flick(icon_deny, src)
-				return
+			try_vend(key, user)
+	if(.)
+		add_fingerprint(user)
 
-			vend_ready = FALSE // From this point onwards, vendor is locked to performing this transaction only, until it is resolved.
 
-			if(!ishuman(user) || R.price <= 0)
-				// Either the purchaser is not human, or the item is free.
-				// Skip all payment logic.
-				vend(R, user)
-				add_fingerprint(user)
-				vend_ready = TRUE
-				. = TRUE
-				return
+/obj/machinery/economy/vending/proc/try_vend(key, mob/user)
+	if(!vend_ready)
+		to_chat(user, "<span class='warning'>The vending machine is busy!</span>")
+		return
+	if(panel_open)
+		to_chat(user, "<span class='warning'>The vending machine cannot dispense products while its service panel is open!</span>")
+		return
+
+	var/list/display_records = product_records + coin_records
+	if(extended_inventory)
+		display_records = product_records + coin_records + hidden_records
+	if(key < 1 || key > length(display_records))
+		to_chat(user, "<span class='warning'>ERROR: invalid inum passed to vendor. Report this bug.</span>")
+		return
+	var/datum/data/vending_product/R = display_records[key]
+	if(!istype(R))
+		to_chat(user, "<span class='warning'>ERROR: unknown vending_product record. Report this bug.</span>")
+		return
+	var/list/record_to_check = product_records + coin_records
+	if(extended_inventory)
+		record_to_check = product_records + coin_records + hidden_records
+	if(!R || !istype(R) || !R.product_path)
+		to_chat(user, "<span class='warning'>ERROR: unknown product record. Report this bug.</span>")
+		return
+	if(R in hidden_records)
+		if(!extended_inventory)
+			// Exploit prevention, stop the user purchasing hidden stuff if they haven't hacked the machine.
+			to_chat(user, "<span class='warning'>ERROR: machine does not allow extended_inventory in current state. Report this bug.</span>")
+			return
```

Comment:
these should throw runtimes. not tell them to report it as a bug.

also you dont need to do if(R) `istype(R)` will return 0 if R is null, and you already check for istype(R) on line 629.

Reply:
redundant istypes removed, to chats are converted to debug logs